worker/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+    mut child: Child,
+    inner: Arc<ProcessInner>,
+    stop_receiver: oneshot::Receiver<()>,
+) {
+    // A child process might close both stdout and stderr,
+    // but remain alive. In that case, we must still try to wait
+    // for the stop message.
+    // Fuse the future to be able to .await twice safely (when !is_terminated()).
+    let mut stop_receiver = stop_receiver.fuse();
+
+    let stdout = BufReader::new(child.stdout.take().expect("should always be available"));
+    let stderr = BufReader::new(child.stderr.take().expect("should always be available"));
+
+    // Phase 1: copy logs from stdout/stderr, on stop message: signal the child.
+    tokio::select! {
+        copied = futures::future::join(
+            logs::copy(stdout, inner.clone()),
+            logs::copy(stderr, inner.clone())
+        ) => {
+            if let Err(e) = copied.0 {
+                error!("{:?}", e);
+            }
+            if let Err(e) = copied.1 {
+                error!("{:?}", e);
+            }
+        },
+        _ = &mut stop_receiver => {
+            if let Err(e) = ops::stop_child(&mut child).await {
+                error!("{:?}", e);
+            }
+        },
+    };
+
+    // Phase 2: stdout/stderr have been closed,
+    // wait for a stop message to arrive (if not arrived yet),
+    // or on the child to finish otherwise
+    loop {
+        tokio::select! {
+            _ = &mut stop_receiver, if !stop_receiver.is_terminated() => {
+                if let Err(e) = ops::stop_child(&mut child).await {
+                    error!("{:?}", e);
+                }
+            },
+            res = child.wait() => {
+                match res {
+                    Ok(s) => {
+                        inner.finish(s).await;
+                        inner.progress.notify_waiters();
+                        return;
+                    }
+                    Err(_) => unimplemented!("when does this happen?"), // TODO
+                }
+            }
+        }
+    }
+}
+
+pub(crate) struct ProcessInner {
+    exit_status: RwLock<Option<ExitStatus>>,
+    logs: RwLock<Vec<Bytes>>,
+
+    // Signals the listeners about progress being made by the process
+    // (either new log messages or finishing).
+    progress: Arc<Notify>,
+
+    stop_sender: Mutex<Option<oneshot::Sender<()>>>,
+}
+
+impl ProcessInner {
+    fn new(stop_sender: oneshot::Sender<()>) -> Self {
+        Self {
+            exit_status: Default::default(),
+            logs: Default::default(),
+            progress: Default::default(),
+            stop_sender: Mutex::new(Some(stop_sender)),
+        }
+    }
+    async fn finish(&self, exit_status: ExitStatus) {
+        *self.exit_status.write().await = Some(exit_status);
+    }
+}
+
+/// Represents a single process.
+pub struct Process(Arc<ProcessInner>);
+
+impl Process {
+    /// Spawn a new process.
+    pub fn spawn<'a>(argv0: &str, argv: impl Iterator<Item = &'a str>) -> Result<Self, io::Error> {
```

Comment:
I believe the lifetimes here can be elided and newer versions of clips should also nag about it.

Reply:
Hmm, does not seem to be the case with Rust 1.50.0

```
error[E0106]: missing lifetime specifier
   --> worker/src/lib.rs:107:58
    |
107 |     pub fn spawn(argv0: &str, argv: impl Iterator<Item = &str>) -> Result<Self, io::Error> {
    |                                                          ^ expected named lifetime parameter
    |
help: consider introducing a named lifetime parameter
    |
107 |     pub fn spawn<'a>(argv0: &str, argv: impl Iterator<Item = &'a str>) -> Result<Self, io::Error> {
    |                 ^^^^                                         ^^^
```

In fact, that's what I tried at first and modified per compiler recommendations.

Reply:
Ah, that's something that has changed recently then. Since this is the recommendation from rust, consider this resolved.